interfaces/builtin/adb_support.go

```diff
@@ -0,0 +1,214 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+const adbSupportSummary = `allows access to connected USB devices for use with fastboot or adb`
+
+const adbSupportBaseDeclarationSlots = `
+  adb-support:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+// These udev rules are specifically for making it convenient for people
+// to connect to their device without requiring root privileges. Upstream
+// adb also uses 'GROUP="plugdev"', but this group is not guaranteed
+// to exist and since MODE is 0666, it isn't required. We use 'MODE="0666"'
+// since the access to the USB devices doesn't provide a privilege
+// escalation on the host and the connected Android device has its own
+// checks for allowing access that are independent of the host USB
+// setup, therefore a more restricted setup (eg 'MODE="0660"') doesn't
+// appreciably improve security.
+var adbSupportConnectedPlugUDev = []string{
+	// ACER
+	`SUBSYSTEM=="usb", ATTR{idVendor}=="0502", MODE="0666"`,
```

Comment:
If we consider two commands in a snap or two snaps that are installed that plugs adb-support, these MODE assignment rules don't like right to me because they will be replicated for each command. I think we instead want these MODE rules in adbSupportPermanentSlotUDev (please verify that does what you want as an implicit slot whose rules might be influenced by re-exec).

Then the adbSupportConnectedPlugUDev becomes:
```
var adbSupportConnectedPlugUDev = []string{
	// ACER
	`SUBSYSTEM=="usb", ATTR{idVendor}=="0502"`,
	// ALLWINNER
	`SUBSYSTEM=="usb", ATTR{idVendor}=="1f3a"`,
...
}
```

In this manner, the mode rules are added only once and the assignment rules are added in the usual way.

Note, I'm not super excited that we are adding these udev MODE rules unconditionally to the system as a part of snapd being on the system. If we instead used the slot app approach I suggested, this would be avoided since they would only be put on the system when a snap providing the adb implementation was installed, which is least surprising.